src/sensor_icm20948.c

```diff
@@ -103,12 +105,30 @@ ic20948_query(struct icm20948 *ic, uint8_t oid)
     // If we have enough bytes to fill the buffer do it and send report
     if (ic->fifo_pkts_bytes >= BYTES_PER_BLOCK) {
         uint8_t reg = AR_FIFO;
-        read_mpu(ic->i2c, sizeof(reg), &reg, BYTES_PER_BLOCK, &ic->sb.data[0]);
-        ic->sb.data_count = BYTES_PER_BLOCK;
+        uint8_t *d = &ic->sb.data[ic->sb.data_count];
+        memset(d, 0, BYTES_PER_BLOCK);
+        int r = read_mpu(ic->i2c, sizeof(reg), &reg,
+                         BYTES_PER_BLOCK, d);
+        if (r) {
+            if (r == I2C_BUS_START_NACK || r == I2C_BUS_START_READ_NACK)
+                goto out;
+            int i = 0;
+            for (; i < BYTES_PER_BLOCK; i++) {
+                if (d[i] != 0)
+                    break;
+            }
+            if (i == BYTES_PER_BLOCK)
+                goto out;
+            // Unaligned read
+            i2c_shutdown_on_err(r);
+        }
+        ic->sb.data_count += BYTES_PER_BLOCK;
         ic->fifo_pkts_bytes -= BYTES_PER_BLOCK;
-        sensor_bulk_report(&ic->sb, oid);
+        if (ic->sb.data_count + BYTES_PER_BLOCK > ARRAY_SIZE(ic->sb.data))
+            sensor_bulk_report(&ic->sb, oid);
```

Comment:
This looks odd to me.  I think that if there is a transfer error we should discard the entire transfer and increase the overflow counter (`ic->sb.possible_overflows++;`).  I don't think the code should attempt to use a partial transfer - that seems error prone.

Reply:
This code does almost that. I'm not sure it is even possible to handle partial reads.
So, I just ensure there is no read happened at all, and reading of those bytes will be continued in the next cycle.

If partial reads just happened, it still triggers the `i2c_shutdown_on_err(r);` here.

Btw, possible_overflows seems not to be used by the code above, also it is a little optimistic for ICM and MPU.
Because overflow can happen more frequently than 10 times per second.
But looks like it is only checked below, by the host query (once per 0.1s).